File: ruby/lib/epostak/resources/integrator.rb
# frozen_string_literal: true

module EPostak
  module Resources
    # Integrator-aggregate endpoints (+sk_int_*+ keys only).
    #
    # Access via +client.integrator.licenses.info+. For the per-firm
    # +/account+ and +/licenses/info+ views (which integrators also reach via
    # +X-Firm-Id+), use +client.account+ instead.
    #
    # @example
    #   usage = client.integrator.licenses.info(limit: 100)
    #   puts usage["billable"]["totalCharge"], "EUR this month"
    class Integrator
      # @return [Resources::IntegratorKeys] Integrator API-key management.
      attr_reader :keys

      # @return [Resources::IntegratorLicenses] Billing aggregate views.
      attr_reader :licenses

      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @keys = IntegratorKeys.new(http)
        @licenses = IntegratorLicenses.new(http)
      end
    end

    # +/integrator/keys+ -- integrator API-key management.
    class IntegratorKeys
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # List all API keys for the current integrator.
      #
      # @return [Hash] Response with +keys+
      def list
        @http.request(:get, "/integrator/keys")
      end

      # Deactivate an integrator API key by UUID or +sk_int_*+ prefix.
      #
      # @param key_id [String, nil] API key UUID
      # @param client_id [String, nil] +sk_int_*+ key prefix/client ID
      # @return [Hash] Success response
      def deactivate(key_id: nil, client_id: nil)
        body = {}
        body[:keyId] = key_id unless key_id.nil?
        body[:client_id] = client_id unless client_id.nil?
        @http.request(:delete, "/integrator/keys", body: body)
      end
    end

    # +/integrator/licenses/*+ -- billing aggregate views.
    #
    # Tier rates are applied to the AGGREGATE document count across all the
    # integrator's +integrator-managed+ firms. A 100-firm x 50-doc integrator
    # lands in tier 2-3, not tier 1 like a standalone firm would. Volumes
    # above +contactThreshold+ (5 000 / month) flip +exceedsAutoTier+ to
    # +true+; auto-billing pauses there and sales handles invoicing manually.
    class IntegratorLicenses
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # Aggregate plan + current-period usage across managed firms.
      #
      # Wraps +GET /api/v1/integrator/licenses/info+. Requires the
      # +account:read+ scope on a +sk_int_*+ integrator key. No +X-Firm-Id+
      # header -- the endpoint is integrator-scoped, not firm-scoped.
      #
      # @param offset [Integer] Pagination offset for the per-firm list
      #   (default 0).
      # @param limit [Integer] Page size for the per-firm list, max 100
      #   (default 50).
      # @return [Hash] Response with +integrator+, +period+, +nextResetAt+,
      #   +billable+ (managed-plan aggregate + tier-applied charges),
      #   +nonManaged+, +exceedsAutoTier+, +contactThreshold+,
      #   +pricing+ ({+outboundTiers+, +inboundApiTiers+}), paginated +firms+
      #   rows, and +pagination+.
      # @raise [EPostak::Error] On API error
      #
      # @example
      #   usage = client.integrator.licenses.info(limit: 100)
      #   if usage["exceedsAutoTier"]
      #     # sales review required, auto-billing has paused
      #   end
      def info(offset: 0, limit: 50)
        @http.request(:get, "/integrator/licenses/info", query: { offset: offset, limit: limit })
      end
    end
  end
end

File: python/src/epostak/resources/integrator.py
"""Integrator-aggregate resource (``sk_int_*`` keys only).

Exposes endpoints that operate at the integrator level (no ``X-Firm-Id``).
For per-firm ``/account`` and ``/licenses/info`` views (which integrators
also reach via ``X-Firm-Id``), use :mod:`epostak.resources.account` instead.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, Optional

import httpx

if TYPE_CHECKING:
    from epostak.types import IntegratorLicenseInfo

from epostak.resources.documents import _BaseResource, _build_query


class IntegratorLicensesResource(_BaseResource):
    """``/integrator/licenses/*`` — billing aggregate views."""

    def info(
        self,
        offset: int = 0,
        limit: int = 50,
    ) -> IntegratorLicenseInfo:
        """Aggregate plan + current-period usage across managed firms.

        Wraps ``GET /api/v1/integrator/licenses/info``. Tier rates are applied
        to the AGGREGATE document count (not per-firm summed) — a 100-firm ×
        50-doc integrator lands in tier 2-3, not tier 1 like a standalone
        firm would.

        Volumes above ``contactThreshold`` (5 000 / month) flip
        ``exceedsAutoTier`` to ``True``; auto-billing pauses there and sales
        handles invoicing manually.

        Args:
            offset: Pagination offset for the per-firm list (default 0).
            limit: Page size for the per-firm list, max 100 (default 50).

        Returns:
            Dict with ``integrator``, ``period``, ``nextResetAt``,
            ``billable`` (managed-plan aggregate + tier-applied charges),
            ``nonManaged`` (firms paying their own plan),
            ``exceedsAutoTier``, ``contactThreshold``, ``pricing``,
            ``firms`` (paginated per-firm rows), and ``pagination``.

        Requires the ``account:read`` scope on a ``sk_int_*`` integrator key.
        No ``X-Firm-Id`` header — the endpoint is integrator-scoped.

        Example::

            usage = client.integrator.licenses.info(limit=100)
            if usage["exceedsAutoTier"]:
                # sales review required, auto-billing has paused
                ...
            print(usage["billable"]["totalCharge"], "EUR this month")
        """
        params = _build_query({"offset": offset, "limit": limit})
        return self._request("GET", "/integrator/licenses/info", params=params)


class IntegratorResource:
    """Integrator-aggregate endpoints (``sk_int_*`` only)."""

    def __init__(
        self,
        client: httpx.Client,
        base_url: str,
        token_manager: object,
        firm_id: Optional[str],
        *,
        max_retries: int = 3,
        _rate_limit_store: Optional[list] = None,
    ) -> None:
        self.licenses = IntegratorLicensesResource(
            client, base_url, token_manager, firm_id, max_retries=max_retries,
            _rate_limit_store=_rate_limit_store,
        )
